### PR TITLE
Introduce Storyline editor

### PR DESCRIPTION
**Pull request by mgrenonville**
introduce jointjs as graph library
Introduce Storyline container
Select adgroup instead of campaign

https://github.com/MEDIARITHMICS/mediarithmics-api/pull/225
 https://github.com/MEDIARITHMICS/core-platform-services/pull/162

### CODE REVIEW

**Comment by thomas-tosoni** [comment]
Pull Request #200 !

**Comment by dduponchel** [comment]
You need to update the bower.json file: `joint` is not a declared dependency. The build also fails with a missing jsPlumb 2.0.6 dependency.

File: app/src/core/scenarios/EditOneController.js [hunk 1]
```diff
@@ -139,7 +163,14 @@ define(['./module'], function (module) {
         }
         promise.then(function success(campaignContainer) {
           $log.info("success");
-          $location.path('/' + Session.getCurrentWorkspace().organisation_id + "/library/scenarios");
+          if(!$scope.graph.scenarioId) {
+            $scope.graph.saveWithNewScenario(campaignContainer);
```
Comment (dduponchel) [inline app/src/core/scenarios/EditOneController.js]: You should flatMap the promises.

File: app/src/core/scenarios/JointJsDirectives.js [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+   */
+  module.directive('jsJointEndpoint', function ($log) {
+    var def = {
+      restrict: 'E',
+      require: '^jsJointCanvas',
+      scope: {
+        settings: '=settings'
+      },
+      link: function (scope, element, attrs, jsJointCanvas) {
+        var instance = jsJointCanvas.scope.jsPlumbInstance;
```
Comment (dduponchel) [inline app/src/core/scenarios/JointJsDirectives.js]: `jsPlumbInstance` looks like a bad copy/paste.

File: app/src/core/scenarios/PlumbDirective.js [hunk 3]
```diff
@@ -1,201 +0,0 @@
...
-        // });
-      }
-    };
-    return def;
-  });
-
-  /**
-   * JS PLUMB OBJECT
-   */
-  module.directive('jsPlumbObject', function () {
```
Comment (dduponchel) [inline app/src/core/scenarios/PlumbDirective.js]: This directive is still used in `app/src/core/scenarios/partials/defaultState.html` and `app/src/core/scenarios/partials/secondaryState.html`.

File: app/src/core/scenarios/StorylineContainer.js [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+          return deferred.promise;
+        };
+        this.saveWithNewScenario = function (newScenario) {
+          this.scenario = newScenario;
+          var self = this;
+          this.storyline = newScenario.one("storyline").get().then(function (storyline) {
+            self.storyline = storyline;
+            _.forEach(self.nodes, function (node){ node.scenario = newScenario;});
+            _.forEach(self.edges, function (edge){ edge.scenario = newScenario;});
+            self.save();
```
Comment (dduponchel) [inline app/src/core/scenarios/StorylineContainer.js]: You should `return` this line to flatMap promises.

File: app/src/core/scenarios/StorylineContainer.js [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+          });
+
+        };
+
+        this.save = function save() {
+          var deferred = $q.defer();
+          var self = this;
+          self.saveNodes(self.nodes, "save").then(function () {
```
Comment (dduponchel) [inline app/src/core/scenarios/StorylineContainer.js]: Flattening promises give better readability:
```js
return self.saveNodes(self.nodes, "save")
.then(function () {
  return self.saveNodes(self.edges, "save");
}).then(function () {
  return self.saveNodes(self.removedEdges, "delete");
}).then(function () {
  return self.saveNodes(self.removedNodes, "delete");
}).then(function () {
  return self;
});
```

Comment (mgrenonville) [inline app/src/core/scenarios/StorylineContainer.js]: [Indeed...](http://www.flatmapthatshit.com/)

File: app/src/core/scenarios/PlumbDirective.js (hunk 3 above)
Comment (mgrenonville) [inline app/src/core/scenarios/PlumbDirective.js]: DONE !

**Pull request merged**